### PR TITLE
br: add more options for br sql client

### LINKED ISSUE #53040: Add more options for br sql
## Feature Request

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

We need add the following options for br sql client:

the general options are:

`ENCRYPTION_METHOD` for Encrypt/decrypt method, be one of plaintext|aes128-ctr|aes192-ctr|aes256-ctr case-insensitively, "plaintext" represents no encrypt/decrypt (default "plaintext")

`ENCRYPTION_KEY_FILE` for FilePath, its content is used as the cipher-key

`CHECKSUM_CONCURRENCY` for The concurrency of checksumming in one table (default 4)

the options for backup are :

`IGNORE_STATS` for wether ignore backup stats (default true)

`COMPRESSION_LEVEL` for compression level used for sst file compression

`BACKUP_COMPRESSION` for backup sst file compression algorithm, value can be one of 'lz4|zstd|snappy' (default "zstd")

the options for restore are:

`WAIT_TIFLASH_READY` for whether wait tiflash replica ready if tiflash exists

`WITH_SYS_TABLE` for whether restore system privilege tables on default setting (default true)

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by RidRisR**
### What problem does this PR solve?

Issue Number: close #53040 

Problem Summary:

### What changed and how does it work?

We added the following options for br sql client:

the general options are:

`ENCRYPTION_METHOD` for Encrypt/decrypt method, be one of plaintext|aes128-ctr|aes192-ctr|aes256-ctr case-insensitively, "plaintext" represents no encrypt/decrypt (default "plaintext")

`ENCRYPTION_KEYFILE` for FilePath, its content is used as the cipher-key

`CHECKSUM_CONCURRENCY` for The concurrency of checksumming in one table (default 4)

the options for backup are :

`IGNORE_STATS` for wether ignore backup stats (default true)

`COMPRESSION_LEVEL` for compression level used for sst file compression

`COMPRESSION_TYPE` for backup sst file compression algorithm, value can be one of 'lz4|zstd|snappy' (default "zstd")

the options for restore are:

`WAIT_TIFLASH_READY` for whether wait tiflash replica ready if tiflash exists

`WITH_SYS_TABLE` for whether restore system privilege tables on default setting (default true)

`LOAD_STATS` for whether run load stats at end of snapshot restore task (default true)

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
We added the following options for br sql client:

the general options are:

`ENCRYPTION_METHOD` for Encrypt/decrypt method, be one of plaintext|aes128-ctr|aes192-ctr|aes256-ctr case-insensitively, "plaintext" represents no encrypt/decrypt (default "plaintext")

`ENCRYPTION_KEYFILE` for FilePath, its content is used as the cipher-key

`CHECKSUM_CONCURRENCY` for The concurrency of checksumming in one table (default 4)

the options for backup are :

`IGNORE_STATS` for wether ignore backup stats (default true)

`COMPRESSION_LEVEL` for compression level used for sst file compression

`COMPRESSION_TYPE` for backup sst file compression algorithm, value can be one of 'lz4|zstd|snappy' (default "zstd")

the options for restore are:

`WAIT_TIFLASH_READY` for whether wait tiflash replica ready if tiflash exists

`WITH_SYS_TABLE` for whether restore system privilege tables on default setting (default true)

`LOAD_STATS` for whether run load stats at end of snapshot restore task (default true)```

### CODE REVIEW

File: pkg/executor/brie.go [hunk 1]
```diff
@@ -391,8 +426,16 @@ func (b *executorBuilder) buildBRIE(s *ast.BRIEStmt, schema *expression.Schema)
 		rcfg.Config = cfg
 		e.restoreCfg = &rcfg
 		for _, opt := range s.Options {
-			if opt.Tp == ast.BRIEOptionOnline {
+			// if opt.Tp == ast.BRIEOptionOnline {
+			// 	e.restoreCfg.Online = opt.UintValue != 0
+			// }
```
Comment (YuJuncen) [inline pkg/executor/brie.go:431]:
```suggestion
```

File: pkg/executor/brie.go [hunk 2]
```diff
@@ -383,6 +402,22 @@ func (b *executorBuilder) buildBRIE(s *ast.BRIEStmt, schema *expression.Schema)
...
+			case ast.BRIEOptionCompression:
+				switch opt.StrValue {
+				case "zstd":
+					e.backupCfg.CompressionConfig.CompressionType = backuppb.CompressionType_ZSTD
+				case "snappy":
+					e.backupCfg.CompressionConfig.CompressionType = backuppb.CompressionType_SNAPPY
+				case "lz4":
+					e.backupCfg.CompressionConfig.CompressionType = backuppb.CompressionType_LZ4
+				default:
+					e.backupCfg.CompressionConfig.CompressionType = backuppb.CompressionType_ZSTD
```
Comment (YuJuncen) [inline pkg/executor/brie.go:414]: Perhaps keep unmodified compression type when not found?

File: pkg/executor/brie.go [hunk 3]
```diff
@@ -330,6 +333,22 @@ func (b *executorBuilder) buildBRIE(s *ast.BRIEStmt, schema *expression.Schema)
...
+			case "aes128-ctr":
+				cfg.CipherInfo.CipherType = encryptionpb.EncryptionMethod_AES128_CTR
+			case "aes192-ctr":
+				cfg.CipherInfo.CipherType = encryptionpb.EncryptionMethod_AES192_CTR
+			case "aes256-ctr":
+				cfg.CipherInfo.CipherType = encryptionpb.EncryptionMethod_AES256_CTR
+			case "plaintext":
+				cfg.CipherInfo.CipherType = encryptionpb.EncryptionMethod_PLAINTEXT
+			default:
+				cfg.CipherInfo.CipherType = encryptionpb.EncryptionMethod_PLAINTEXT
```
Comment (YuJuncen) [inline pkg/executor/brie.go:349]: Perhaps keep it unmodified when not found?

File: pkg/parser/keywords.go [hunk 4]
```diff
@@ -353,6 +356,8 @@ var Keywords = []KeywordsType{
 	{"ENABLE", false, "unreserved"},
 	{"ENABLED", false, "unreserved"},
 	{"ENCRYPTION", false, "unreserved"},
+	{"ENCRYPTION_KEY_FILE", false, "unreserved"},
```
Comment (YuJuncen) [inline pkg/parser/keywords.go:359]: It seems this option is an no-op. Could you add some test cases about this?

File: pkg/parser/keywords.go [hunk 5]
```diff
@@ -619,10 +625,12 @@ var Keywords = []KeywordsType{
 	{"VIEW", false, "unreserved"},
 	{"VISIBLE", false, "unreserved"},
 	{"WAIT", false, "unreserved"},
+	{"WAIT_TIFLASH_READY", false, "unreserved"},
```
Comment (YuJuncen) [inline pkg/parser/keywords.go:628]: ...and this. I guess you may check all options you added in the test case?

File: pkg/executor/brie.go (hunk 3 above)
Comment (RidRisR) [reply pkg/executor/brie.go:356]: ok, if then it should return nil.

File: pkg/executor/brie.go (hunk 2 above)
Comment (RidRisR) [reply pkg/executor/brie.go:421]: ok, if then it should return nil.

File: pkg/executor/brie_test.go [hunk 6]
```diff
@@ -140,3 +144,82 @@ func TestFetchShowBRIE(t *testing.T) {
...
+		case ast.BRIEOptionLoadStats:
+			require.True(t, opt.UintValue == 1)
+		}
+	}
+	schema = plan.Schema()
+	exec = builder.buildBRIE(s, schema)
+	require.NoError(t, builder.err)
+	e, ok = exec.(*BRIEExec)
+	require.True(t, ok)
+	require.True(t, e.restoreCfg.ChecksumConcurrency == 4)
```
Comment (YuJuncen) [inline pkg/executor/brie_test.go:221]: Preferring use `require.Equal`, which will show the values when assertion fails.
```suggestion
	require.Equal(t, e.restoreCfg.ChecksumConcurrency, 4)
```

File: pkg/executor/brie_test.go [hunk 7]
```diff
@@ -140,3 +144,82 @@ func TestFetchShowBRIE(t *testing.T) {
...
+	exec := builder.buildBRIE(s, schema)
+	require.NoError(t, builder.err)
+	e, ok := exec.(*BRIEExec)
+	require.True(t, ok)
+	require.True(t, e.backupCfg.ChecksumConcurrency == 4)
+	require.True(t, e.backupCfg.CompressionLevel == 4)
+	require.True(t, e.backupCfg.IgnoreStats)
+	require.True(t, e.backupCfg.CompressionConfig.CompressionType == backuppb.CompressionType_LZ4)
+	require.True(t, e.backupCfg.CipherInfo.CipherType == encryptionpb.EncryptionMethod_AES256_CTR)
+	require.True(t, len(e.backupCfg.CipherInfo.CipherKey) > 0)
```
Comment (YuJuncen) [inline pkg/executor/brie_test.go:195]: Preferring use `require.Greater`.

**Review by YuJuncen — APPROVED** [review]
rest lgtm

**Review by Leavrth — APPROVED** [review]

**Pull request merged**